drivers/bm/bm_dfu_core.c

```diff
@@ -0,0 +1,342 @@
+ * Copyright (c) 2015 Intel Corporation
+ * Copyright (c) 2022 Sofar Ocean Technologies
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <init.h>
+#include <logging/log.h>
+#include <storage/flash_map.h>
+#include <bootutil/bootutil_public.h>
+#include <dfu/mcuboot.h>
+#include <sys/reboot.h>
+
+#include <drivers/bm/bm_serial.h>
+#include <drivers/bm/bm_dfu.h>
+#include <drivers/bm/bm_dfu_client.h>
+#include <drivers/bm/bm_dfu_host.h>
+#include <drivers/bm/bm_common.h>
+#include <kernel.h>
+
+LOG_MODULE_REGISTER(bm_dfu, CONFIG_BM_LOG_LEVEL);
+
+static struct k_thread _subsystem_dfu_thread_data;
+static struct k_thread _transport_service_dfu_thread_data;
+
+K_THREAD_STACK_DEFINE(_bm_dfu_subsystem_stack, CONFIG_BM_DFU_TASK_STACK_SIZE);
+K_MSGQ_DEFINE(_dfu_subsystem_queue, sizeof(bm_dfu_event_t), CONFIG_BM_DFU_NUM_FRAMES, 4);
+
+K_THREAD_STACK_DEFINE(_bm_dfu_transport_service_stack, CONFIG_BM_DFU_TASK_STACK_SIZE);
+K_MSGQ_DEFINE(_dfu_transport_service_queue, sizeof(bm_msg_t), CONFIG_BM_DFU_NUM_FRAMES, 4);
+
+static const struct smf_state dfu_states[];
+static dfu_core_ctx_t _dfu_context;
+
+/* ******************************************************************* */
+/* ********** State machine entry, run and exit functions ************ */
+/* ******************************************************************* */
+
+static void s_init_run(void *o)
+{
+    if (_dfu_context.current_event.type == DFU_EVENT_INIT_SUCCESS)
+    {
+        smf_set_state(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_IDLE]);
+    }
+}
+
+static void s_idle_run(void *o)
+{
+    if (_dfu_context.current_event.type == DFU_EVENT_RECEIVED_UPDATE_REQUEST)
+    {
+#ifdef CONFIG_BM_DFU_CLIENT
+        /* Client */
+        bm_dfu_client_process_request();
+#endif
+    }
+    else if (_dfu_context.current_event.type == DFU_EVENT_BEGIN_UPDATE)
+    {
+#ifdef CONFIG_BM_DFU_HOST
+        /* Host */
+        smf_set_state(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_HOST_REQ_UPDATE]);
+#endif
+    }
+}
+
+static void s_error_entry(void *o)
+{
+    /* TODO: What do we do here? */
+    LOG_ERR("Entered the Error State.");
+
+    switch (_dfu_context.error)
+    {
+        case BM_DFU_ERR_FLASH_ACCESS:
+            LOG_ERR("Flash access error (Fatal Error)");
+            break;
+        case BM_DFU_ERR_IMG_CHUNK_ACCESS:
+            LOG_ERR("Unable to get image chunk");
+        case BM_DFU_ERR_NONE:
+        case BM_DFU_ERR_TOO_LARGE:
+            LOG_ERR("Image too large for Client");
+        case BM_DFU_ERR_SAME_VER:
+            LOG_ERR("Client already loaded with image");
+        case BM_DFU_ERR_MISMATCH_LEN:
+            LOG_ERR("Length mismatch");
+        case BM_DFU_ERR_BAD_CRC:
+            LOG_ERR("CRC mismatch");
+        case BM_DFU_ERR_TIMEOUT:
+            LOG_ERR("DFU Timeout");
+        case BM_DFU_ERR_BM_FRAME:
+            LOG_ERR("BM Processing Error");
+        default:
+            smf_set_state(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_IDLE]);
```

Comment:
Should these all fall through, or are we missing `break;` statements

Reply:
I have all but one case fall through to the default case where we just move back to the IDLE state. The only case that has a break is the Flash write error, which we've considered to be "Fatal"

Reply:
If BM_DFU_ERR_IMG_CHUNK_ACCESS happens, you're going to see a lot of confusing LOG_ERR's

Reply:
Oh duh. Should I just go ahead and duplicate the "set_state" and break for each case?

Reply:
You could just create an errorcode-to-string utility method to retrieve the appropriate log string and make a single LOG_ERR() call with that result.